packages/astro/src/core/app/index.ts

```diff
@@ -398,21 +402,25 @@ export class App {
 			return newResponse;
 		}
 
-		const { statusText, headers } = oldResponse;
-
 		// If the new response did not have a meaningful status, an override may have been provided
 		// If the original status was 200 (default), override it with the new status (probably 404 or 500)
 		// Otherwise, the user set a specific status while rendering and we should respect that one
 		const status = override?.status
 			? override.status
-			: oldResponse.status === 200
+			: originalResponse.status === 200
 			  ? newResponse.status
-			  : oldResponse.status;
+			  : originalResponse.status;
 
 		return new Response(newResponse.body, {
 			status,
-			statusText: status === 200 ? newResponse.statusText : statusText,
-			headers: new Headers(Array.from(headers)),
+			statusText: status === 200 ? newResponse.statusText : originalResponse.statusText,
+			// If you're looking at here for possible bugs, it means that it's not a bug.
+			// With the middleware, users can meddle with headers, and we should pass to the 404/500.
+			// If users see something weird, it's because they are setting some headers they should not.
+			headers: new Headers([
+				...Array.from(newResponse.headers),
+				...Array.from(originalResponse.headers),
+			]),
```

Comment:
This will have an odd effect when the originalResponse is `new Response("whatever string", { status: "404" })`

Reply:
What do you suggest to do?

Reply:
I think the solution would be to remove `Content-type` from originalResponse.

We are replacing content, it's natural we replace content type too.

Reply:
Done, I also updated the test case where we attempt to change the content type header